### ISSUE TITLE
windowEndpoint smells bad.

### ISSUE THREAD

**Issue by PaulKinlan** [GoogleChrome/comlink#6]
WindowEndpoint smells bad. Why does it exist? Why can't you just detect that its the endpoint and manage it. As a consumer of the endpoint I should never have to care.

**Comment by surma** [comment]
I was going back and forth between this – here’s my reasoning:

Currently, `Endpoint` is defined as follows:

```typescript
export interface Endpoint {
    postMessage(message: any, transfer?: any[]): void;
    addEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: {}): void;
    removeEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: {}): void;
}
```

It’s an object similar to a `MessageChannel`. Something that you can put messages into and something that you can receive messages on.

A window, however, is special. Because you want to call `postMessage()` on the window object, but want to call `{add,remove}EventListener` on something else – mostly `self`. 

I chose the explicit solution. 

Happy to discuss this, obv.

**Comment by PaulKinlan** [comment]
Just to nit pick, wrt Endpoint you meant similar to `MessagePort` and not `MessageChannel` and when that is the case, that endpoint interface looks almost exactly like the window interface outside of a the origin filter in postMessage.

Right now I have to do windowEndpoint on both ends if either the opener or opened is an iframe or a window. Ideally the user of comlink shouldn't care, it should do it all for the user of the API and work it out and I just don't see the issue that this is trying to solve.

**Comment by surma** [comment]
Lemme break it down, maybe you have a better idea to solve this:

The main difference is that with windows you want to _receive_ messages on a different window object than the one you are _sending_ on. I could indeed automatically detect that, but I thought making it explicit would be preferable.

Implementing the `Endpoint` interface on top of `MessagePort`:

```js
Comlink.proxy({
  postMessage: port.postMessage.bind(port),
  addEventListener: port.addEventListener.bind(port),
  removeEventListener: port.removeEventListener.bind(port),
});
```

Implementing the `Endpoint` interface on top of `Window`:

```js
Comlink.proxy({
  postMessage: someWindow.postMessage.bind(someWindow),
  addEventListener: self.addEventListener.bind(self),
  removeEventListener: self.removeEventListener.bind(self),
});
```

**Comment by surma** [comment]
... the more I think about it the more I think you are right.

**Comment by surma** [comment]
Fixed in 190f16fc83d91f0728328b9be4e180bb644bcebd
